### PR TITLE
Correct pointer(::AbstractArray, i) to use strides, deprecate elsize(::AbstractArray)

### PR DESCRIPTION
**Pull request by mbauman**
Previously, `pointer` was erroneously calculating memory offsets based upon `linearindices` instead of `strides` and `Base.elsize`.  This corrects it.  This is a bug-fix for StridedArrays that did not have contiguous strides, and arrays that do not specialize `strides` will see the same behavior but with a deprecation warning that the fallback `strides` implementation (that assumes contiguous strides) is deprecated.

Edit: I've now also deprecated the fallback `elsize` instead of coercing `Array`'s definition onto everyone.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
I'm pretty sure there are places where I've written code that assumes that `sizeof(a) == length(a)*sizeof(eltype(a))`. Can we just eliminate the special cases and make that true?

**Comment by mbauman** [comment]
We simply don't know how `a` is going to arrange its elements in memory.  It could do snazzy things like Array's bitsunion splitting, or it could handle some other datatypes in special ways.  Jameson spitballed a good example last night:

```julia
struct TaggedArray{T,N} <: AbstractArray{T, N}
    data::Array{Tuple{T, Int},N}
end
getindex(A::TaggedArray, i) = A.data[i][1]
specialsauce(A::TaggedArray, i) = A.data[i][2]
strides(A::TaggedArray) = strides(A.data)
```

Now `pointer(::TaggedArray, ::Int)` is going to be wrong for this type.  It cannot even account for this with its `strides` definition, since `sizeof(Tuple{T, Int})/sizeof(T)` may not be an integer (consider `T=Int128`).

I think `elsize` needs to go the way of `strides` — we must deprecate this fallback for `AbstractArray`s and require strided arrays to define it themselves.  I'll push that change in a bit.

**Comment by vtjnash** [comment]
> (consider T=Int128)

Alignment considerations will just manage to save you there. My favorite to reach for is thus often something like `T = NTuple{7, UInt8}`

**Comment by mbauman** [comment]
Alright, I've changed things around here to use `Base.elsize` much more consistently and carefully.

**Comment by mbauman** [comment]
Hrm, I can't seem to reproduce the segfault locally, but I'm changing around `pointer` definitions… so… it's worrisome.

**Comment by vchuravy** [comment]
One other thing to consider are compressed arrays like `BitArray`. I am not quite sure how you would implement `memoffset` for those types.

**Comment by mbauman** [comment]
Easy — we don't.  They're not strided arrays.

File: base/abstractarray.jl [hunk 1]
```diff
@@ -122,7 +122,13 @@ prevind(::AbstractArray, i::Integer) = Int(i)-1
 nextind(::AbstractArray, i::Integer) = Int(i)+1
 
 eltype(::Type{<:AbstractArray{E}}) where {E} = @isdefined(E) ? E : Any
-elsize(::AbstractArray{T}) where {T} = sizeof(T)
+
+"""
+    Base.elsize(::Type{<:CustomArray})
```
Comment (JeffBezanson) [inline base/abstractarray.jl]: Doesn't match the signature of the documented method.

**Comment by mbauman** [comment]
I should know better than to couple so many changes together.  Splitting this apart.

**Pull request closed without merging**